thriftbp/circuit_breaker.go

```diff
@@ -0,0 +1,102 @@
+package thriftbp
+
+import (
+	"context"
+	"time"
+
+	"github.com/apache/thrift/lib/go/thrift"
+	"github.com/reddit/baseplate.go/log"
+	"github.com/reddit/baseplate.go/metricsbp"
+	"github.com/sony/gobreaker"
+)
+
+// FailureRatioBreaker is a circuit breaker based on gobreaker that uses a low-water-mark and
+// % failure threshold to trip.
+type FailureRatioBreaker struct {
+	*gobreaker.CircuitBreaker
+
+	metricPrefix      string
+	minRequestsToTrip int
+	failureThreshold  float64
+}
+
+// CircuitBreakerConfig represents the configuration for a FailureRatioBreaker
+type CircuitBreakerConfig struct {
+	// Minimum requests that need to be sent during a time period before the breaker is eligible to transition from closed to open.
+	MinRequestsToTrip int
+	// Percentage of requests that need to fail during a time period for the breaker to transition from closed to open.
+	// Represented as a float in {0,1}, where .05 means >5% failures will trip the breaker.
+	FailureThreshold float64
+	// Gobreaker settings used to configure the underlying circuit breaker.
+	Settings gobreaker.Settings
+}
+
+// NewFailureRatioBreaker creates a new FailureRatioBreaker with the provided configuration.
+func NewFailureRatioBreaker(config CircuitBreakerConfig) FailureRatioBreaker {
+
+	failureBreaker := FailureRatioBreaker{
+		metricPrefix:      config.Settings.Name,
+		minRequestsToTrip: config.MinRequestsToTrip,
+		failureThreshold:  config.FailureThreshold,
+	}
+	config.Settings.OnStateChange = stateChanged
+	config.Settings.ReadyToTrip = failureBreaker.ShouldTripCircuitBreaker
+	failureBreaker.CircuitBreaker = gobreaker.NewCircuitBreaker(config.Settings)
+	go failureBreaker.runStatsProducer()
+	return failureBreaker
+}
+
+func (cb FailureRatioBreaker) runStatsProducer() {
+	circuitBreakerGauge := metricsbp.M.RuntimeGauge(cb.metricPrefix + ".circuit.breaker.closed")
+
+	tick := time.NewTicker(10 * time.Second)
+	defer tick.Stop()
+	for {
+		select {
+		case <-metricsbp.M.Ctx().Done():
+			return
+		case <-tick.C:
+			if cb.State() == gobreaker.StateOpen {
+				circuitBreakerGauge.Set(0)
+			} else {
+				circuitBreakerGauge.Set(1)
+			}
+		}
+	}
+}
+
+// ThriftMiddleware is a thrift.ClientMiddleware that handles circuit breaking.
+func (cb FailureRatioBreaker) ThriftMiddleware(next thrift.TClient) thrift.TClient {
```

Comment:
This is inside `thriftbp` package so the name `ThriftMiddleware` is kinda redundant. Make rename it into `ClientMiddleware` or just `Middleware` (as it doesn't make sense as a server middleware).

Reply:
Maybe this should instead be in its own package? Then `ThriftClientMiddleware` makes sense (and also is cleaner to add new middleware's over time).

Reply:
I'm kind of on the fence on this one. This only makes sense as a client middleware, and currently we don't have any http client middleware framework, so thrift client is the only one, thus it makes sense to live under `thriftbp` for now, and move to its own package in the future when we really need that (as compare to retry this is a thinner wrapper around a third party library). But then if we really move it to its own package in the future that would be a breaking change.

Reply:
Ya, I kinda prefer having it be in its own package to not set up a potential breaking change but I don’t feel super strongly.

Reply:
just `breakerbp`?